pam_sss_gss: support authentication indicators

abbra opened the pull request:
MIT Kerberos allows to associate authentication indicators with the
issued ticket based on the way how the TGT was obtained. The indicators
present in the TGT then copied to service tickets. There are two ways to
check the authentication indicators:

 - when KDC issues a service ticket, a policy at KDC side can reject the
   ticket issuance based on a lack of certain indicator

 - when a server application presented with a service ticket from a
   client, it can verify that this ticket contains intended
   authentication indicators before authorizing access from the client.

Add support to validate presence of a specific (set of) authentication
indicator(s) in pam_sss_gss when validating a user's TGT.

This concept can be used to only allow access to a PAM service when user
is in possession of a ticket obtained using some of pre-authentication
mechanisms that require multiple factors: smart-cards (PKINIT), 2FA
tokens (otp/radius), etc.

Resolves: https://github.com/SSSD/sssd/issues/5482
Signed-off-by: Alexander Bokovoy <abokovoy@redhat.com>

abbra commented:
This is a draft implementation of https://github.com/SSSD/sssd/issues/5482. I haven't tested it yet myself.

abbra commented:
It now works for me. Here is a test on Fedora 33:
```
[admin@master ~]$ export KRB5CCNAME=/tmp/admin.cc
[admin@master ~]$ sudo -l
pam_sss_gss: Initializing GSSAPI authentication with SSSD
pam_sss_gss: Switching euid from 0 to 1169800000
pam_sss_gss: Trying to establish security context
pam_sss_gss: SSSD User name: admin@ipa.test
pam_sss_gss: User domain: ipa.test
pam_sss_gss: User principal: 
pam_sss_gss: Target name: host@master.ipa.test
pam_sss_gss: Using ccache: /tmp/admin.cc
pam_sss_gss: Acquiring credentials, principal name will be derived
pam_sss_gss: Switching euid from 1169800000 to 0
pam_sss_gss: Authentication successful
Matching Defaults entries for admin on master:
    !visiblepw, always_set_home, match_group_by_gid, always_query_group_plugin, env_reset, env_keep="COLORS DISPLAY HOSTNAME HISTSIZE KDEDIR LS_COLORS", env_keep+="MAIL QTDIR USERNAME LANG LC_ADDRESS LC_CTYPE", env_keep+="LC_COLLATE LC_IDENTIFICATION
    LC_MEASUREMENT LC_MESSAGES", env_keep+="LC_MONETARY LC_NAME LC_NUMERIC LC_PAPER LC_TELEPHONE", env_keep+="LC_TIME LC_ALL LANGUAGE LINGUAS _XKB_CHARSET XAUTHORITY",
    secure_path=/usr/local/sbin\:/usr/local/bin\:/usr/sbin\:/usr/bin\:/sbin\:/bin\:/var/lib/snapd/snap/bin

User admin may run the following commands on master:
    (root) ALL
[admin@master ~]$ logout
[root@master ~]# grep pam_gssapi /etc/sssd/sssd.conf
pam_gssapi_services = sudo, sudo-i
pam_gssapi_indicators_map = hardened, sudo:pkinit, sudo-i:otp
[root@master ~]# fgrep gssapi_ /var/log/sssd/sssd_pam.log |tail -10
(2021-02-08 17:18:50): [pam] [pam_cmd_gssapi_init_done] (0x0400): Trying GSSAPI auth: User[admin@ipa.test], Domain[ipa.test], UPN[], Target[host@master.ipa.test]
(2021-02-08 17:18:50): [pam] [pam_cmd_gssapi_init_done] (0x0400): Returning [0]: Success
(2021-02-08 17:18:50): [pam] [gssapi_handshake] (0x0400): Security context established with [admin@IPA.TEST]
(2021-02-08 17:18:50): [pam] [gssapi_get_indicators] (0x0400): attribute's [auth-indicators] value [hardened] authenticated
(2021-02-08 17:18:50): [pam] [gssapi_get_indicators] (0x0400): authentication indicators: [hardened]
(2021-02-08 17:18:50): [pam] [pam_gssapi_check_indicators] (0x0400): indicator [hardened] is allowed for PAM service [sudo]
(2021-02-08 17:18:50): [pam] [pam_cmd_gssapi_sec_ctx] (0x0400): Check if acquired service ticket has req. indicators: 0
(2021-02-08 17:18:50): [pam] [pam_cmd_gssapi_sec_ctx] (0x0400): Checking that target user matches UPN
(2021-02-08 17:18:50): [pam] [pam_cmd_gssapi_sec_ctx_done] (0x0400): User [admin@ipa.test] match UPN [admin@IPA.TEST]. Authentication was successful.
(2021-02-08 17:18:50): [pam] [pam_cmd_gssapi_sec_ctx_done] (0x0400): Returning [0]: Success

```

frozencemetery commented on src/responder/pam/pamsrv_gssapi.c:
```diff
@@ -385,12 +469,130 @@ static char *gssapi_get_name(TALLOC_CTX *mem_ctx, gss_name_t gss_name)
+    if (exported == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE,
+              "Unable to pre-allocate indicators\n");
+        return NULL;
```
`attrs` leaks here.  I suggest using a single exit path - i.e., `goto done` for early exit since there are several allocated objects that need to be handled in this function (attrs, value, display_value, exported).

abbra replied:
value/display_value defined within the scope of the internal loop, so for common `goto done;` style they'd require moving things out to the function scope. This should be OK, I think. I'll do it.

frozencemetery commented on src/responder/pam/pamsrv_gssapi.c:
```diff
@@ -385,12 +469,130 @@ static char *gssapi_get_name(TALLOC_CTX *mem_ctx, gss_name_t gss_name)
+    }
+
+    for (int i = 0; i < attrs->count; i++) {
+        gss_buffer_desc value;
+        gss_buffer_desc display_value;
```
These need to be initialized.

frozencemetery commented on src/responder/pam/pamsrv_gssapi.c:
```diff
@@ -385,12 +469,130 @@ static char *gssapi_get_name(TALLOC_CTX *mem_ctx, gss_name_t gss_name)
+            if (major != GSS_S_COMPLETE) {
+                DEBUG(SSSDBG_CRIT_FAILURE,
+                        "Unable to retrieve an attribute\n");
+                gss_release_buffer(&minor, &value);
+                gss_release_buffer(&minor, &display_value);
```
These don't need to be handled on failure.  As long as they're initialized above, though, this is harmless (but right now it's not).

frozencemetery commented on src/responder/pam/pamsrv_gssapi.c:
```diff
@@ -385,12 +469,130 @@ static char *gssapi_get_name(TALLOC_CTX *mem_ctx, gss_name_t gss_name)
+                (void)gss_release_buffer_set(&minor, &attrs);
+                return NULL;
+            }
+            (void)gss_release_buffer(&minor, &value);
+            (void)gss_release_buffer(&minor, &display_value);
```
While I think it would be better to use `goto done` cleanup for this function as a whole, if you choose not to do that, these two lines can go before the `if` above.

frozencemetery commented on src/responder/pam/pamsrv_gssapi.c:
```diff
@@ -385,12 +469,130 @@ static char *gssapi_get_name(TALLOC_CTX *mem_ctx, gss_name_t gss_name)
+
+    (void)gss_release_buffer_set(&minor, &attrs);
+
+    if ((exported != NULL) && (exported[0] != '\0')) {
```
Won't `exported` always be non-NULL here?

frozencemetery reviewed (changes requested):
Nothing seems majorly amiss, but some stuff in comments.

I have reviewed for GSSAPI correctness and general C correctness.  I have not reviewed for talloc-related stuff, or (due to time constraints) project style.

While there are not tests here, I'm not familiar enough with sssd's testing infrastructure to say how difficult they would be to add.  (If a KDC is being set up already, spake is probably the easiest indicator to check.)

abbra commented on src/man/sssd.conf.5.xml:
```diff
@@ -1771,6 +1771,50 @@ pam_gssapi_services = sudo, sudo-i
+                            </itemizedlist>
+                        </para>
+                        <para>
+                            Example: to require access to SUDO services only for users which obtained their Kerberos tickets with a smart card-based pre-authentication (PKINIT), set
```
Correct. It is an example here. Adding 'or any other certificate-based authentication' is stretching it too much with the details.

abbra commented:
Thanks, @frozencemetery, I updated the code according to the suggestions.

I also decided to unify a bit the man page example language -- I do talk about x.509-based certificates when describing authentication indicators in `sssd.conf(5)` right before the example where it is stated that these certificates can be stored in files or on smart cards.

tiran commented on src/man/sssd.conf.5.xml:
```diff
@@ -1771,6 +1771,59 @@ pam_gssapi_services = sudo, sudo-i
+                           pam_sss_gss.so module.
+                        </para>
+                        <para>
+			   Each element of the list can be either authentication indicator
+                           name or a pair <quote>service:indicator</quote>. Indicators not
+			   prefixed with the PAM service name will be required to be present
+                           to access any PAM service configured to be used with
+                           <option>pam_gssapi_services</option>.
```
It's not clear to me if items in the list of indicator are *match any* or *match all*. For example does ``pam_gssapi_indicators_map = sudo:pkinit, sudo:otp, hardened`` require either

* one or more indicators of `pkinit`, `otp`, and `hardened`
* all three indicators
* either `(pkinit or otp) and hardened`

abbra replied:
This option works together with the `pam_gssapi_services` which defines list of PAM services where `pam_sss_gss.so` would be allowed for use. `pam_gsssapi_indicators_map` provides a list of indicators that must be present when performing authentication for those PAM services. It is enough to have one of the listed indicators in the ticket. However, because people might want to have different settings for different PAM services, there is a way to set that per-PAM service with a prefix `service:`. To avoid repeating the same values, non-prefixed indicator applies to all services in `pam_gssapi_services`.

In your example, if `pam_gssapi_services = sudo` and `pam_gssapi_indicators_map = sudo:pkinit, sudo:otp, hardened`, when we'd do a check for `sudo` PAM service, then presence of either `pkinit`, `otp`, or `hardened` would allow to pass the check.

When `pam_sss_gss.so` applies authentication indicators check, the code will build a list of indicators that apply to the specific PAM service at the point of the check. Kerberos tickets typically have only one indicator, produced by the pre-authentication method used for obtaining TGT, so the code only checks that any of the indicators in the ticket match the list of indicators associated with the PAM service.

I don't think a presence of multiple indicators at the same time should be checked. I certainly don't see how we can express that in SSSD configuration without introducing a special parser. There might be no practical reason for that (yet?).

In practice, for IPA environments we only assign a single indicator during pre-authentication method processing. We folded up both FAST and SPAKE into the same `hardened` authentication indicator. We might add a recognition of MS-PKCA-obtained Kerberos tickets and inject `pkinit` indicator to service tickets in IPA for the users from Active Directory who used a smartcard to authenticate to their Windows machines. However, I don't see a need for the checks of multiple indicators at the same time.

Upstream, MIT Kerberos has internal tests that assign more than one indicator and verify that a received ticket has them all. So far it looks like a theoretical construct rather than a practical need. Being able to validate indicators against a list of values is more useful.

tiran replied:
Thanks for your clarification. I agree that an "match any" approach makes the most sense to me.

To me, the documentation of the ``pam_gssapi_indicators_map`` option does not make it obvious that it's a "match any" option.

abbra replied:
I'll update it.

abbra replied:
@tiran I extended man pages, hopefully, it is now more clear.

frozencemetery reviewed (approved):
Thanks!  I'm happy now (but see caveats in previous review, and we wait for tiran before merging).

sumit-bose commented on src/responder/pam/pamsrv_gssapi.c:
```diff
@@ -18,12 +18,14 @@
     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
 
+#include <asm-generic/errno-base.h>
```
Hi,

why is this `#include` needed? I do not get a compilation error if I remove it and imo it is already include via `errno.h`.

bye,
Sumit

abbra replied:
I don't know. It looks like an overly smart IDE I tried to use for this PR preparation. ;) Fixed.

sumit-bose commented on src/responder/pam/pamsrv_gssapi.c:
```diff
@@ -83,6 +85,102 @@ static bool pam_gssapi_should_check_upn(struct pam_ctx *pam_ctx,
     return pam_ctx->gssapi_check_upn;
 }
 
+static int pam_gssapi_check_indicators(struct pam_ctx *pam_ctx,
```
Hi,

please add a comment which explains the special meaning of `EINVAL` with respect to access control. 

`domain` is only used to assign `map`, why not just pass `map` directly?

`pam_ctx` is a long living context which will be freed only if the process exits. `cli_ctx` from the caller might be more suitable. I think a generic `TALLOC_CTX *mem_ctx` would fit better here.

bye,
Sumit

sumit-bose replied:
Hi,

after reading the function more carefully I wonder if `EINVAL` should really be treated as access allowed as suggested by the comment in the caller. Do you by chance mean `ENOENT` ?

bye,
Sumit

abbra replied:
An idea was to differentiate cases when internal processing fails and we really cannot decide whether it is a broken configuration. The comment in the caller should have really said ENOENT, indeed. What is a typical pattern in SSSD for the situations when you cannot decide whether a broken setup should allow or deny access? I'd probably tend to use denial here.

alexey-tikhonov commented:
For some reason covscan fails to build:
(f32) https://cov01.lab.eng.brq.redhat.com/covscanhub/task/202028/log/stdout.log
(f33) https://cov01.lab.eng.brq.redhat.com/covscanhub/task/202031/log/stdout.log
(sorry for internal links)

Most probably issue isn't related to the patch, as "vanilla" 2.4.1/2.4.2 fails to build the same way (https://cov01.lab.eng.brq.redhat.com/covscanhub/task/202034/log/stdout.log), but this blocks covscan as a part of review process.

sumit-bose commented on src/man/sssd.conf.5.xml:
```diff
@@ -1771,6 +1771,65 @@ pam_gssapi_services = sudo, sudo-i
+                           Note: This option can also be set per-domain which
+                           overwrites the value in [pam] section. It can also
+                           be set for trusted domain which overwrites the value
+                           in the domain section.
```
Hi,

the PR currently only implements reading the option from the specific domain or sub-domain section. Using the option in the [pam] section in not implemented and also inheriting to trusted domains not implemented.

bye,
Sumit

abbra replied:
Thanks! I implemented them too.

abbra commented:
@alexey-tikhonov I fixed couple covscan-reported issues. There are others which are due to covscan not finishing its analysis in the limited time and thus killing its child process by timeout.

abbra commented:
@alexey-tikhonov I ran covscan for my patches and while it still failed, at least two issues addressed by the separate patches here were fixed. I seem few more, though. They are the same as in 2.4.0/2.4.1 runs, so auth-indicators support didn't add anything new.

abbra commented:
Added few more covscan fixes.

@sumit-bose asked on IRC to add a way to 'cancel' authentication indicator check in a subdomain. I added support for `-` to achieve that, so it can be specified in a `pam_gssapi_indicators_map` in a subdomain configuration directly or per PAM service, e.g. `service:-`.

abbra commented:
@alexey-tikhonov so I looked closer to covscan failure that prevents the clean run. It is due to the following problem:
```
RPM build errors:
error: Empty %files file /builddir/build/BUILD/sssd-2.4.2/sssd_client.lang
    Empty %files file /builddir/build/BUILD/sssd-2.4.2/sssd_client.lang
[WARNING] Build command sh -c unset XML_CATALOG_FILES; sh -c "rpmbuild -bi --nodeps --short-circuit /builddir/build/SPECS/sssd.spec '--define' '_without_testsuite 1' '--define' 'apidocs 0' '--define' 'libguestfs_runtests 0' '--define' 'runselftest 0' '--define' 'with_publican 0' '--without' 'binfilter' '--without' 'docs' '--without' 'langpacks' '--define' '_smp_mflags -j13' '--nocheck'" exited with code 1. Please verify that the build completed successfully.
[WARNING] No files were emitted. This may be due to a problem with your configuration
or because no files were actually compiled by your build command.
Please make sure you have configured the compilers actually used in the compilation.
 For more details, please look at: 
    /builddir/cov/build-log.txt
Finish: chroot ['/bin/su mockbuild -l -c "PATH=/usr/lib64/csclng:/usr/lib64/cscppc:/usr/lib64/csgcca:/usr/lib64/cswrap:\\$PATH CSWRAP_DEL_CXXFLAGS=\\$\'-Werror*:-fdiagnostics-color*:-no-canonical-prefixes:-Wno-error=deprecated-register\' CSWRAP_DEL_CFLAGS=\\$\'-Werror*:-fdiagnostics-color*:-no-canonical-prefixes:-Wno-error=deprecated-register\' CSWRAP_ADD_CFLAGS=\\$\'-Wno-unknown-pragmas\' CSWRAP_ADD_CXXFLAGS=\\$\'\' CSWRAP_TIMEOUT_FOR=\\$\'::cppcheck:clang:clang++:gcc\' COVERITY_UNSUPPORTED_COMPILER_INVOCATION=\\$\'1\' CSWRAP_TIMEOUT=\\$\'30\' CSWRAP_CAP_FILE=\\$\'/builddir/cswrap-capture.err\' sh -c \\"/opt/coverity/bin/cov-build --dir /builddir/cov --append-log sh -c \\\\\\"unset XML_CATALOG_FILES; sh -c \\\\\\\\\\\\\\"rpmbuild -bi --nodeps --short-circuit /builddir/build/SPECS/sssd.spec \'--define\' \'_without_testsuite 1\' \'--define\' \'apidocs 0\' \'--define\' \'libguestfs_runtests 0\' \'--define\' \'runselftest 0\' \'--define\' \'with_publican 0\' \'--without\' \'binfilter\' \'--without\' \'docs\' \'--without\' \'langpacks\' \'--define\' \'_smp_mflags -j13\' \'--nocheck\'\\\\\\\\\\\\\\"\\\\\\"\\""']

!!! 2021-02-10 09:35:08	error: %install failed

```
I saw this issue myself and reported it to @pbrezina already. This is definitely a difference between 2.4.1 release and current git master because when I am adding the patches from this PR to 2.4.1, there is no failure in rebuilds.

alexey-tikhonov commented:
Hi @abbra,

> I saw this issue myself and reported it to @pbrezina already. This is definitely a difference between 2.4.1 release and current git master because when I am adding the patches from this PR to 2.4.1, there is no failure in rebuilds.

Yes, it seems our change of specfile triggered an odd behavior of '--nocheck' option covscan uses.

There is similar BZ: https://bugzilla.redhat.com/1907766

I'm trying to move '-%check' section around spec-file to see if it helps but no luck so far.

alexey-tikhonov commented:
I used `+%define _empty_manifest_terminate_build 0` to work-around covscan build issue and run a `version-diff-build`.
I can confirm that this PR doesn't bring in any new covscan warning but rather fixes 24 old covscan warnings.
I'm not sure if patches fixing covscan issues in the code unrelated to main purpose of PR should belong here though... but anyway, thanks a lot @abbra!

pbrezina commented:
> I used `+%define _empty_manifest_terminate_build 0` to work-around covscan build issue and run a `version-diff-build`.
> I can confirm that this PR doesn't bring in any new covscan warning but rather fixes 24 old covscan warnings.
> I'm not sure if patches fixing covscan issues in the code unrelated to main purpose of PR should belong here though... but anyway, thanks a lot @abbra!

Would using `%find_lang` macro help?

alexey-tikhonov commented:
> Would using `%find_lang` macro help?

I don't know, probably. But lets discuss this in a dedicated email thread as this has nothing to do with this PR and will be lost here...

abbra commented:
I'll move covscan fixes to a separate PR.

abbra commented:
Moved the covscan fixes to PR https://github.com/SSSD/sssd/pull/5498

sumit-bose commented:
Hi,

thanks for the update. In my testing I only found that inheriting the option from configured to trusted domain does not work. Something like:

```
diff --git a/src/db/sysdb_subdomains.c b/src/db/sysdb_subdomains.c
index 675d41c..e2381c8 100644
--- a/src/db/sysdb_subdomains.c
+++ b/src/db/sysdb_subdomains.c
@@ -191,6 +191,7 @@ struct sss_domain_info *new_subdomain(TALLOC_CTX *mem_ctx,
     dom->override_gid = parent->override_gid;
 
     dom->gssapi_services = parent->gssapi_services;
+    dom->gssapi_indicators_map = parent->gssapi_indicators_map;
 
     if (parent->sysdb == NULL) {
         DEBUG(SSSDBG_OP_FAILURE, "Missing sysdb context in parent domain.\n");
```

should fix it.

bye,
Sumit

abbra commented:
@sumit-bose thank you! I added this one line and re-pushed.

sumit-bose commented:
Thanks, ACK.

bye,
Sumit

pbrezina commented:
Pushed PR: https://github.com/SSSD/sssd/pull/5494

* `master`
    * 5ce7ced269c7b3dd8f75122a50f539083b5697ae - pam_sss_gss: support authentication indicators
